akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,256 @@
+import akka.util.Timeout
+
+import scala.concurrent.duration._
+import scala.collection.immutable
+import scala.concurrent.Await
+import scala.reflect.ClassTag
+
+/**
+ * Messages for subscribing to changes in the cluster state
+ *
+ * Not intended for user extension.
+ */
+@DoNotInherit
+sealed trait ClusterStateSubscription
+
+object Subscribe {
+  def apply[A <: ClusterDomainEvent: ClassTag](subscriber: ActorRef[A]) =
+    new Subscribe[A, A](subscriber, implicitly[ClassTag[A]].runtimeClass.asInstanceOf[Class[A]])
+}
+
+/**
+ * Subscribe to cluster state changes. The initial state of the cluster will be sent as
+ * a "replay" of the subscribed events.
+ *
+ * If all you are interested in is to react on this node becoming a part of the cluster,
+ * see [[OnSelfUp]]
+ *
+ * @param subscriber A subscriber that will receive events until it is unsubscribed or stops
+ * @param eventClass The type of events to subscribe to, can be individual event types such as
+ *                   `ReachabilityEvent` or one of the common supertypes, such as `MemberEvent` to get
+ *                   all the subtypes of events.
+ */
+final case class Subscribe[A <: ClusterDomainEvent, S <: A](
+  subscriber: ActorRef[A],
+  eventClass: Class[S]) extends ClusterStateSubscription
+
+/**
+ * Subscribe to this node being up, after sending this event the subscription is automatically
+ * cancelled. If the node is already up the event is also sent to the subscriber.
+ */
+final case class OnSelfUp(subscriber: ActorRef[SelfUp]) extends ClusterStateSubscription
+
+/**
+ * @param currentClusterState The cluster state snapshot from when the node became Up.
+ */
+final case class SelfUp(currentClusterState: CurrentClusterState)
+
+final case class Unsubscribe[T](subscriber: ActorRef[T]) extends ClusterStateSubscription
+final case class GetCurrentState(recipient: ActorRef[CurrentClusterState]) extends ClusterStateSubscription
+
+/**
+ * Not intended for user extension.
+ */
+@DoNotInherit
+sealed trait ClusterCommand
+
+/**
+ * Try to join this cluster node with the node specified by 'address'.
+ *
+ * An actor system can only join a cluster once. Additional attempts will be ignored.
+ * When it has successfully joined it must be restarted to be able to join another
+ * cluster or to join the same cluster again.
+ *
+ * The name of the [[akka.actor.ActorSystem]] must be the same for all members of a
+ * cluster.
+ */
+final case class Join(address: Address) extends ClusterCommand
+
+/**
+ * Join the specified seed nodes without defining them in config.
+ * Especially useful from tests when Addresses are unknown before startup time.
+ *
+ * An actor system can only join a cluster once. Additional attempts will be ignored.
+ * When it has successfully joined it must be restarted to be able to join another
+ * cluster or to join the same cluster again.
+ */
+final case class JoinSeedNodes(seedNodes: immutable.Seq[Address]) extends ClusterCommand
```

Comment:
do we use `Util.immutableSeq` for the Java API, or should we have a separate factory method?